src/D2L.Bmx/OktaAuthenticator.cs

```diff
@@ -114,26 +119,142 @@ No config file found. Your Okta session will not be cached.
+				if( Array.Find( cookies, c => c.Name == "sid" )?.Value is string sid ) {
+					sessionIdTcs.SetResult( sid );
+				}
+			}
+		} catch( TaskCanceledException ) {
+			consoleWriter.WriteWarning( $"""
+				WARNING: Timed out when trying to create Okta session through Desktop Single Sign-On.
+				Check if the org '{orgUrl}' is correct. If running BMX with elevated privileges,
+				rerun the command with the '--experimental-bypass-browser-security' flag
+				"""
+			);
+			return null;
+		} catch( TargetClosedException ) {
+			consoleWriter.WriteWarning( """
+				WARNING: Failed to create Okta session through Desktop Single Sign-On as BMX is likely being run
+				with elevated privileges. Rerun the command with the '--experimental-bypass-browser-security' flag.
+				"""
+			);
+			return null;
+		} catch( Exception ) {
+			consoleWriter.WriteWarning(
+				"WARNING: Unknown error while trying to authenticate with Okta using Desktop Single Sign-On." );
+			return null;
+		}
+
+		if( sessionId is null ) {
+			return null;
+		}
+
+		var oktaAuthenticatedClient = oktaClientFactory.CreateAuthenticatedClient( orgUrl, sessionId );
+		var oktaSession = await oktaAuthenticatedClient.GetCurrentOktaSessionAsync();
+		if( !OktaUserMatchesProvided( oktaSession.Login, user ) ) {
+			consoleWriter.WriteWarning(
+				"WARNING: Could not create Okta session using Desktop Single Sign-On as provided Okta user "
+				+ $"'{StripLoginDomain( user )}' does not match user '{StripLoginDomain( oktaSession.Login )}'." );
+			return null;
+		}
+
+		TryCacheOktaSession( user, orgUrl.Host, sessionId, oktaSession.ExpiresAt );
+		return oktaAuthenticatedClient;
+	}
+
+	private static string StripLoginDomain( string email ) {
+		return email.Contains( '@' ) ? email.Split( '@' )[0] : email;
+	}
+
+	private static bool OktaUserMatchesProvided( string oktaLogin, string providedUser ) {
+		string adName = StripLoginDomain( oktaLogin );
+		string normalizedUser = StripLoginDomain( providedUser );
+		return adName.Equals( normalizedUser, StringComparison.OrdinalIgnoreCase );
+	}
+
+	private bool TryCacheOktaSession( string userId, string org, string sessionId, DateTimeOffset expiresAt ) {
+		if( File.Exists( BmxPaths.CONFIG_FILE_NAME ) ) {
+			CacheOktaSession( userId, org, sessionId, expiresAt );
+			return true;
+		}
+		consoleWriter.WriteWarning( """
+					No config file found. Your Okta session will not be cached.
+					Consider running `bmx configure` if you own this machine.
+					""" );
+		return false;
+	}
+
 	private void CacheOktaSession( string userId, string org, string sessionId, DateTimeOffset expiresAt ) {
 		var session = new OktaSessionCache( userId, org, sessionId, expiresAt );
 		var sessionsToCache = ReadOktaSessionCacheFile();
-		sessionsToCache = sessionsToCache.Where( session => session.UserId != userId && session.Org != org )
+		sessionsToCache = sessionsToCache.Where( session => session.UserId != userId || session.Org != org )
```

Comment:
Spotted we don't support caching same username for multiple orgs. Did we want this?

Reply:
don't think it was intentional